File: src/main/java/me/shedaniel/linkie/commands/YarnMethodCommand.java
package me.shedaniel.linkie.commands;

import me.shedaniel.linkie.CommandBase;
import me.shedaniel.linkie.InvalidUsageException;
import me.shedaniel.linkie.LinkieBot;
import me.shedaniel.linkie.yarn.MappingsData;
import me.shedaniel.linkie.yarn.YarnManager;
import me.shedaniel.linkie.yarn.YarnMethod;
import net.fabricmc.mappings.EntryTriple;
import org.javacord.api.entity.message.MessageAuthor;
import org.javacord.api.entity.message.embed.EmbedBuilder;
import org.javacord.api.event.message.MessageCreateEvent;

import java.awt.*;
import java.util.List;
import java.util.*;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import static me.shedaniel.linkie.LinkieBot.contains;

public class YarnMethodCommand implements CommandBase {
    @Override
    public void execute(ScheduledExecutorService service, MessageCreateEvent event, MessageAuthor author, String cmd, String[] args) {
        if (YarnManager.updating)
            throw new RuntimeException("Yarn is being downloaded, please wait for around 10 seconds");
        if (args.length != 1 && args.length != 2)
            throw new InvalidUsageException("+" + cmd + " <search> [page]");
        event.getChannel().sendMessage(new EmbedBuilder().setTitle("Loading...").setFooter("Requested by " + author.getDiscriminatedName(), author.getAvatar()).setTimestampToNow()).whenComplete((message1, throwable) -> {
            try {
                int page = args.length > 1 ? Integer.parseInt(args[1]) - 1 : 0;
                String name = args[0].indexOf('.') > -1 ? args[0].split("\\.")[1] : args[0];
                String clazz = args[0].indexOf('.') > -1 ? args[0].split("\\.")[0] : null;
                String low = name.toLowerCase(Locale.ROOT);
                String lowClazz = clazz != null ? clazz.toLowerCase(Locale.ROOT) : null;
                List<YarnMethod> files = new ArrayList<>();
                YarnManager.r1_2_5.tiny.getMethodEntries().forEach(methodEntry -> {
                    EntryTriple intermediary = methodEntry.get("intermediary");
                    EntryTriple server = methodEntry.get("server");
                    EntryTriple client = methodEntry.get("client");
                    if (clazz == null || contains(getLast(intermediary.getOwner().toLowerCase(Locale.ROOT)), lowClazz) || (server != null && contains(getLast(server.getOwner().toLowerCase(Locale.ROOT)), lowClazz)) || (client != null && contains(getLast(client.getOwner().toLowerCase(Locale.ROOT)), lowClazz))) {
                        if (contains(getLast(intermediary.getName()).toLowerCase(Locale.ROOT), low) || (server != null && contains(server.getName(), low)) || (client != null && contains(client.getName(), low)))
                            files.add(new YarnMethod(intermediary, server, client));
                    }
                });
                YarnManager.r1_2_5.mappings.forEach(mappingsFile -> {
                    boolean matchClass = clazz == null || contains(getLast(mappingsFile.getIntermediaryClass().toLowerCase(Locale.ROOT)), lowClazz) || contains(getLast(mappingsFile.getYarnClass().toLowerCase(Locale.ROOT)), lowClazz);
                    if (matchClass)
                        for(MappingsData.MethodMappings methodMapping : mappingsFile.getMethodMappings())
                            if (contains(methodMapping.getIntermediaryName().toLowerCase(Locale.ROOT), low) || contains(methodMapping.getYarnName().toLowerCase(Locale.ROOT), low))
                                files.add(new YarnMethod(new EntryTriple(mappingsFile.getIntermediaryClass(), methodMapping.getIntermediaryName(), methodMapping.getIntermediaryDesc()), new EntryTriple(mappingsFile.getYarnClass(), methodMapping.getYarnName(), methodMapping.getIntermediaryDesc())));
                });
                files.forEach(yarnMethod -> {
                    if (yarnMethod.incomplete())
                        if (yarnMethod.needObf()) {
                            YarnManager.r1_2_5.tiny.getMethodEntries().stream().filter(methodEntry -> {
                                return getLast(methodEntry.get("intermediary").getName()).equalsIgnoreCase(getLast(yarnMethod.getIntermediary().getName()));
                            }).findAny().ifPresent(classEntry -> {
                                yarnMethod.setClient(classEntry.get("client"));
                                yarnMethod.setServer(classEntry.get("server"));
                            });
                        } else if (yarnMethod.needMapped()) {
                            for(MappingsData.MappingsFile file : YarnManager.r1_2_5.mappings.stream().filter(mappingsFile -> {
                                return getLast(mappingsFile.getIntermediaryClass()).equalsIgnoreCase(getLast(yarnMethod.getIntermediary().getOwner()));
                            }).collect(Collectors.toList())) {
                                for(MappingsData.MethodMappings methodMapping : file.getMethodMappings()) {
                                    if (methodMapping.getIntermediaryName().equalsIgnoreCase(yarnMethod.getIntermediary().getName())) {
                                        yarnMethod.setMapped(new EntryTriple(file.getYarnClass(), methodMapping.getYarnName(), methodMapping.getIntermediaryDesc()));
                                        break;
                                    }
                                }
                            }
                        }
                });
                List<YarnMethod> cloneFiles = new ArrayList<>(files);
                files.clear();
                cloneFiles.forEach(yarnMethod -> {
                    if (!files.stream().anyMatch(yarnMethod1 -> yarnMethod1.getIntermediary().getName().equals(yarnMethod.getIntermediary().getName())))
                        files.add(yarnMethod);
                });
                if (files.isEmpty())
                    throw new NullPointerException("Match methods not found!");
                files.sort(Comparator.comparingDouble(value -> similarity(get(value, getLast(low)), getLast(low))));
                Collections.reverse(files);
                EmbedBuilder builder = new EmbedBuilder().setTitle("List of Yarn Mappings (Page " + (page + 1) + "/" + (int) Math.ceil(files.size() / 5d) + ")").setFooter("Requested by " + author.getDiscriminatedName(), author.getAvatar()).setTimestampToNow();
                final String[] desc = {""};
                files.stream().skip(page * 5).limit(5).map(yarnMethod -> {
                    if (yarnMethod.needObf())
                        return null;
                    String obf = "client=" + (yarnMethod.getClient() == null ? "null" : yarnMethod.getClient().getName()) + ",server=" + (yarnMethod.getServer() == null ? "null" : yarnMethod.getServer().getName());
                    String main = yarnMethod.getMapped() != null ? yarnMethod.getMapped().getOwner() + "." + yarnMethod.getMapped().getName() : yarnMethod.getIntermediary().getOwner() + "." + yarnMethod.getIntermediary().getName();
                    String methodDesc = (yarnMethod.getMapped() == null ? yarnMethod.getIntermediary().getName() : yarnMethod.getMapped().getName()) + mapDesc(yarnMethod.getIntermediary().getDesc());
                    return "**MC 1.2.5: " + main + "**\n__Name__: " + obf + " => `" + yarnMethod.getIntermediary().getName() + "`" + (yarnMethod.getMapped() != null ? " => `" + yarnMethod.getMapped().getName() + "`" : "") + "\n__Descriptor__: `" + methodDesc + "`\n__Mixin Target__: `" + turnDesc(yarnMethod.getMapped() != null ? yarnMethod.getMapped().getOwner() : yarnMethod.getIntermediary().getOwner()) + methodDesc + "`";
                }).filter(Objects::nonNull).forEach(s -> {
                    if (desc[0].length() + s.length() > 1990)
                        return;
                    if (!desc[0].isEmpty())
                        desc[0] += "\n\n";
                    desc[0] += s;
                });
                builder.setDescription(desc[0].substring(0, Math.min(desc[0].length(), 2000)));
                message1.edit(builder);
                int finalPage[] = {page};
                if (message1.isServerMessage())
                    message1.removeAllReactions().get();
                message1.addReactions("⬅", "❌", "➡").thenRun(() -> {
                    message1.addReactionAddListener(reactionAddEvent -> {
                        try {
                            if (!reactionAddEvent.getUser().getDiscriminatedName().equals(LinkieBot.getApi().getYourself().getDiscriminatedName()) && !reactionAddEvent.getUser().getDiscriminatedName().equals(author.getDiscriminatedName())) {
                                reactionAddEvent.removeReaction();
                                return;
                            }
                            if (reactionAddEvent.getUser().getDiscriminatedName().equals(author.getDiscriminatedName()))
                                if (reactionAddEvent.getEmoji().equalsEmoji("❌")) {
                                    reactionAddEvent.deleteMessage();
                                } else if (reactionAddEvent.getEmoji().equalsEmoji("⬅")) {
                                    reactionAddEvent.removeReaction();
                                    if (finalPage[0] > 0) {
                                        finalPage[0]--;
                                        EmbedBuilder builder1 = new EmbedBuilder().setTitle("List of Yarn Mappings (Page " + (finalPage[0] + 1) + "/" + (int) Math.ceil(files.size() / 5d) + ")").setFooter("Requested by " + author.getDiscriminatedName(), author.getAvatar()).setTimestampToNow();
                                        final String[] desc1 = {""};
                                        files.stream().skip(finalPage[0] * 5).limit(5).map(yarnMethod -> {
                                            if (yarnMethod.needObf())
                                                return null;
                                            String obf = "client=" + (yarnMethod.getClient() == null ? "null" : yarnMethod.getClient().getName()) + ",server=" + (yarnMethod.getServer() == null ? "null" : yarnMethod.getServer().getName());
                                            String main = yarnMethod.getMapped() != null ? yarnMethod.getMapped().getOwner() + "." + yarnMethod.getMapped().getName() : yarnMethod.getIntermediary().getOwner() + "." + yarnMethod.getIntermediary().getName();
                                            String methodDesc = (yarnMethod.getMapped() == null ? yarnMethod.getIntermediary().getName() : yarnMethod.getMapped().getName()) + mapDesc(yarnMethod.getIntermediary().getDesc());
                                            return "**MC 1.2.5: " + main + "**\n__Name__: " + obf + " => `" + yarnMethod.getIntermediary().getName() + "`" + (yarnMethod.getMapped() != null ? " => `" + yarnMethod.getMapped().getName() + "`" : "") + "\n__Descriptor__: `" + methodDesc + "`\n__Mixin Target__: `" + turnDesc(yarnMethod.getMapped() != null ? yarnMethod.getMapped().getOwner() : yarnMethod.getIntermediary().getOwner()) + methodDesc + "`";
                                        }).filter(Objects::nonNull).forEach(s -> {
                                            if (desc1[0].length() + s.length() > 1990)
                                                return;
                                            if (!desc1[0].isEmpty())
                                                desc1[0] += "\n\n";
                                            desc1[0] += s;
                                        });
                                        builder1.setDescription(desc1[0].substring(0, Math.min(desc1[0].length(), 2000)));
                                        message1.edit(builder1);
                                    }
                                } else if (reactionAddEvent.getEmoji().equalsEmoji("➡")) {
                                    reactionAddEvent.removeReaction();
                                    if (finalPage[0] < (int) Math.ceil(files.size() / 5d) - 1) {
                                        finalPage[0]++;
                                        EmbedBuilder builder1 = new EmbedBuilder().setTitle("List of Yarn Mappings (Page " + (finalPage[0] + 1) + "/" + (int) Math.ceil(files.size() / 5d) + ")").setFooter("Requested by " + author.getDiscriminatedName(), author.getAvatar()).setTimestampToNow();
                                        final String[] desc1 = {""};
                                        files.stream().skip(finalPage[0] * 5).limit(5).map(yarnMethod -> {
                                            if (yarnMethod.needObf())
                                                return null;
                                            String obf = "client=" + (yarnMethod.getClient() == null ? "null" : yarnMethod.getClient().getName()) + ",server=" + (yarnMethod.getServer() == null ? "null" : yarnMethod.getServer().getName());
                                            String main = yarnMethod.getMapped() != null ? yarnMethod.getMapped().getOwner() + "." + yarnMethod.getMapped().getName() : yarnMethod.getIntermediary().getOwner() + "." + yarnMethod.getIntermediary().getName();
                                            String methodDesc = (yarnMethod.getMapped() == null ? yarnMethod.getIntermediary().getName() : yarnMethod.getMapped().getName()) + mapDesc(yarnMethod.getIntermediary().getDesc());
                                            return "**MC 1.2.5: " + main + "**\n__Name__: " + obf + " => `" + yarnMethod.getIntermediary().getName() + "`" + (yarnMethod.getMapped() != null ? " => `" + yarnMethod.getMapped().getName() + "`" : "") + "\n__Descriptor__: `" + methodDesc + "`\n__Mixin Target__: `" + turnDesc(yarnMethod.getMapped() != null ? yarnMethod.getMapped().getOwner() : yarnMethod.getIntermediary().getOwner()) + methodDesc + "`";
                                        }).filter(Objects::nonNull).forEach(s -> {
                                            if (desc1[0].length() + s.length() > 1990)
                                                return;
                                            if (!desc1[0].isEmpty())
                                                desc1[0] += "\n\n";
                                            desc1[0] += s;
                                        });
                                        builder1.setDescription(desc1[0].substring(0, Math.min(desc1[0].length(), 2000)));
                                        message1.edit(builder1);
                                    }
                                }
                        } catch (Throwable throwable1) {
                            throwable1.printStackTrace();
                        }
                    }).removeAfter(30, TimeUnit.MINUTES);
                });
            } catch (Throwable throwable1) {
                message1.edit(new EmbedBuilder().setTitle("Linkie Error").setColor(Color.red).setFooter("Requested by " + event.getMessageAuthor().getDiscriminatedName(), event.getMessageAuthor().getAvatar()).addField("Error occurred while processing the command:", throwable1.getClass().getSimpleName() + ": " + throwable1.getLocalizedMessage()).setTimestampToNow());
            }
        });
    }
    
    private String turnDesc(String s) {
        return "L" + s + ";";
    }
    
    private String mapDesc(String desc) {
        String[] split = desc.split("\\)");
        if (split.length > 2)
            throw new IllegalStateException("Wrong Description!");
        String s = split[0];
        if (s.indexOf(';') > -1) {
            for(String aClass : s.substring(1).split(";")) {
                int l = aClass.indexOf("L");
                split[0] = split[0].replace(aClass.substring(l + 1), mapClass(aClass.substring(l + 1)));
            }
        }
        split[1] = mapDescClass(split[1]);
        return String.join(")", split);
    }
    
    private String mapDescClass(String clazz) {
        if (clazz.startsWith("L"))
            return "L" + mapClass(clazz.substring(1, clazz.length() - 1)) + ";";
        if (clazz.startsWith("[[L"))
            return "L" + mapClass(clazz.substring(3, clazz.length() - 1)) + ";";
        return mapClass(clazz);
    }
    
    private String mapClass(String clazz) {
        return YarnManager.r1_2_5.mappings.stream().filter(mappingsFile -> mappingsFile.getIntermediaryClass().equals(clazz)).findAny().map(MappingsData.MappingsFile::getYarnClass).orElse(clazz);
    }
    
    public String get(YarnMethod yarnClass, String search) {
        String intermediary = getLast(yarnClass.getIntermediary().getName());
        if (contains(intermediary, search))
            return intermediary;
        if (yarnClass.getMapped() != null) {
            String mapped = getLast(yarnClass.getMapped().getName()).toLowerCase(Locale.ROOT);
            if (contains(mapped, search))
                return mapped;
        }
        if (yarnClass.getServer() != null) {
            String mapped = getLast(yarnClass.getServer().getName()).toLowerCase(Locale.ROOT);
            if (contains(mapped, search))
                return mapped;
        }
        if (yarnClass.getClient() != null) {
            String mapped = getLast(yarnClass.getClient().getName()).toLowerCase(Locale.ROOT);
            if (contains(mapped, search))
                return mapped;
        }
        return "jdkwjidhwudhuwihduhudwuhiwuhui";
    }
    
    public int editDistance(String s1, String s2) {
        s1 = s1.toLowerCase();
        s2 = s2.toLowerCase();
        
        int[] costs = new int[s2.length() + 1];
        for(int i = 0; i <= s1.length(); i++) {
            int lastValue = i;
            for(int j = 0; j <= s2.length(); j++) {
                if (i == 0)
                    costs[j] = j;
                else {
                    if (j > 0) {
                        int newValue = costs[j - 1];
                        if (s1.charAt(i - 1) != s2.charAt(j - 1))
                            newValue = Math.min(Math.min(newValue, lastValue), costs[j]) + 1;
                        costs[j - 1] = lastValue;
                        lastValue = newValue;
                    }
                }
            }
            if (i > 0)
                costs[s2.length()] = lastValue;
        }
        return costs[s2.length()];
    }
    
    public String getLast(String s) {
        if (s.indexOf('/') <= -1)
            return s;
        String[] s1Split = s.split("/");
        return s1Split[s1Split.length - 1];
    }
    
    public double similarity(String s1, String s2) {
        s1 = getLast(s1);
        s2 = getLast(s2);
        String longer = s1, shorter = s2;
        if (s1.length() < s2.length()) { // longer should always have greater length
            longer = s2;
            shorter = s1;
        }
        int longerLength = longer.length();
        if (longerLength == 0) {
            return 1.0; /* both strings are zero length */
        }
        return (longerLength - editDistance(longer, shorter)) / (double) longerLength;
    }
}

File: src/main/kotlin/me/shedaniel/linkie/commands/YarnMethodCommand.kt
package me.shedaniel.linkie.commands

import discord4j.core.`object`.entity.MessageChannel
import discord4j.core.`object`.entity.User
import discord4j.core.event.domain.message.MessageCreateEvent
import discord4j.core.event.domain.message.ReactionAddEvent
import discord4j.core.spec.EmbedCreateSpec
import me.shedaniel.linkie.*
import me.shedaniel.linkie.utils.dropAndTake
import me.shedaniel.linkie.utils.onlyClass
import me.shedaniel.linkie.utils.similarity
import me.shedaniel.linkie.utils.similarityOnNull
import java.time.Duration
import kotlin.math.ceil
import kotlin.math.min

object YarnMethodCommand : AYarnMethodCommand({ if (it.id.asLong() == 602959845842485258) "1.2.5" else if (it.id.asLong() == 661088839464386571) "1.14.3" else latestYarn }) {
    override fun getName(): String? = "Yarn Method Command"
    override fun getDescription(): String? = "Query yarn methods."
}

object POMFMethodCommand : AYarnMethodCommand({ "b1.7.3" }) {
    override fun getName(): String? = "POMF Method Command"
    override fun getDescription(): String? = "Query pomf methods."
}

open class AYarnMethodCommand(private val defaultVersion: (MessageChannel) -> String, private val isYarn: Boolean = true) : CommandBase {
    override fun execute(event: MessageCreateEvent, user: User, cmd: String, args: Array<String>, channel: MessageChannel) {
        if (args.isEmpty())
            throw InvalidUsageException("!$cmd <search> [version]")
        val mappingsContainerGetter = if (isYarn)
            tryLoadYarnMappingContainer(args.last(), getYarnMappingsContainer(defaultVersion.invoke(channel)))
        else tryLoadMCPMappingContainer(args.last(), getMCPMappingsContainer(defaultVersion.invoke(channel)))
        var searchTerm = args.joinToString(" ").replace('.', '/')
        if (mappingsContainerGetter.first == args.last()) {
            searchTerm = searchTerm.substring(0, searchTerm.lastIndexOf(' '))
        }
        if (searchTerm.contains(' '))
            throw InvalidUsageException("!$cmd <search> [version]")
        val hasClass = searchTerm.contains('/')
        val hasWildcard = (hasClass && searchTerm.substring(0, searchTerm.lastIndexOf('/')).onlyClass() == "*") || searchTerm.onlyClass('/') == "*"

        val message = channel.createEmbed {
            it.apply {
                setFooter("Requested by " + user.discriminatedName, user.avatarUrl)
                setTimestampToNow()
                var desc = "Searching up methods for **${mappingsContainerGetter.first}**."
                if (hasWildcard) desc += "\nCurrently using wildcards, might take a while."
                if (!mappingsContainerGetter.second) desc += "\nThis mappings version is not yet cached, might take some time to download."
                setDescription(desc)
            }
        }.block() ?: throw NullPointerException("Unknown Message!")

        try {
            val mappingsContainer = mappingsContainerGetter.third()

            val classes = mutableMapOf<Class, FindMethodMethod>()
            if (hasClass) {
                val clazzKey = searchTerm.substring(0, searchTerm.lastIndexOf('/')).onlyClass()
                if (clazzKey == "*") {
                    mappingsContainer.classes.forEach { classes[it] = FindMethodMethod.WILDCARD }
                } else {
                    mappingsContainer.classes.forEach { clazz ->
                        when {
                            clazz.intermediaryName.onlyClass().contains(clazzKey, true) -> classes[clazz] = FindMethodMethod.INTERMEDIARY
                            clazz.mappedName?.onlyClass()?.contains(clazzKey, true) == true -> classes[clazz] = FindMethodMethod.MAPPED
                            clazz.obfName.client?.onlyClass()?.contains(clazzKey, true) == true -> classes[clazz] = FindMethodMethod.OBF_CLIENT
                            clazz.obfName.server?.onlyClass()?.contains(clazzKey, true) == true -> classes[clazz] = FindMethodMethod.OBF_SERVER
                            clazz.obfName.merged?.onlyClass()?.contains(clazzKey, true) == true -> classes[clazz] = FindMethodMethod.OBF_MERGED
                        }
                    }
                }
            } else mappingsContainer.classes.forEach { classes[it] = FindMethodMethod.WILDCARD }
            val methods = mutableMapOf<MethodWrapper, FindMethodMethod>()
            val methodKey = searchTerm.onlyClass('/')
            if (methodKey == "*") {
                classes.forEach { (clazz, cm) ->
                    clazz.methods.forEach { method ->
                        if (methods.none { it.key.method == method })
                            methods[MethodWrapper(method, clazz, cm)] = FindMethodMethod.WILDCARD
                    }
                }
            } else {
                classes.forEach { (clazz, cm) ->
                    clazz.methods.forEach { method ->
                        if (methods.none { it.key.method == method })
                            when {
                                method.intermediaryName.contains(methodKey, true) -> methods[MethodWrapper(method, clazz, cm)] = FindMethodMethod.INTERMEDIARY
                                method.mappedName?.contains(methodKey, true) == true -> methods[MethodWrapper(method, clazz, cm)] = FindMethodMethod.MAPPED
                                method.obfName.client?.contains(methodKey, true) == true -> methods[MethodWrapper(method, clazz, cm)] = FindMethodMethod.OBF_CLIENT
                                method.obfName.server?.contains(methodKey, true) == true -> methods[MethodWrapper(method, clazz, cm)] = FindMethodMethod.OBF_SERVER
                                method.obfName.merged?.contains(methodKey, true) == true -> methods[MethodWrapper(method, clazz, cm)] = FindMethodMethod.OBF_MERGED
                            }
                    }
                }
            }
            val sortedMethods = when {
                methodKey == "*" && (!hasClass || searchTerm.substring(0, searchTerm.lastIndexOf('/')).onlyClass() == "*") -> {
                    // Class and method both wildcard
                    methods.entries.sortedBy { it.key.method.intermediaryName }.sortedBy {
                        it.key.parent.mappedName?.onlyClass() ?: it.key.parent.intermediaryName
                    }.map { it.key }
                }
                methodKey == "*" -> {
                    // Only method wildcard
                    val classKey = searchTerm.substring(0, searchTerm.lastIndexOf('/')).onlyClass()
                    methods.entries.sortedBy { it.key.method.intermediaryName }.sortedByDescending {
                        when (it.key.cm) {
                            FindMethodMethod.MAPPED -> it.key.parent.mappedName!!.onlyClass()
                            FindMethodMethod.OBF_CLIENT -> it.key.parent.obfName.client!!.onlyClass()
                            FindMethodMethod.OBF_SERVER -> it.key.parent.obfName.server!!.onlyClass()
                            FindMethodMethod.OBF_MERGED -> it.key.parent.obfName.merged!!.onlyClass()
                            FindMethodMethod.INTERMEDIARY -> it.key.parent.intermediaryName.onlyClass()
                            else -> null
                        }.similarityOnNull(classKey)
                    }.map { it.key }
                }
                hasClass && searchTerm.substring(0, searchTerm.lastIndexOf('/')).onlyClass() != "*" -> {
                    // has class
                    methods.entries.sortedByDescending {
                        when (it.value) {
                            FindMethodMethod.MAPPED -> it.key.method.mappedName!!
                            FindMethodMethod.OBF_CLIENT -> it.key.method.obfName.client!!
                            FindMethodMethod.OBF_SERVER -> it.key.method.obfName.server!!
                            FindMethodMethod.OBF_MERGED -> it.key.method.obfName.merged!!
                            else -> it.key.method.intermediaryName
                        }.onlyClass().similarity(methodKey)
                    }.sortedBy {
                        it.key.parent.mappedName?.onlyClass() ?: it.key.parent.intermediaryName
                    }.map { it.key }
                }
                else -> {
                    methods.entries.sortedBy {
                        it.key.parent.mappedName?.onlyClass() ?: it.key.parent.intermediaryName
                    }.sortedByDescending {
                        when (it.value) {
                            FindMethodMethod.MAPPED -> it.key.method.mappedName!!
                            FindMethodMethod.OBF_CLIENT -> it.key.method.obfName.client!!
                            FindMethodMethod.OBF_SERVER -> it.key.method.obfName.server!!
                            FindMethodMethod.OBF_MERGED -> it.key.method.obfName.merged!!
                            else -> it.key.method.intermediaryName
                        }.onlyClass().similarity(methodKey)
                    }.map { it.key }
                }
            }
            if (sortedMethods.isEmpty())
                throw NullPointerException("No results found!")
            var page = 0
            val maxPage = ceil(sortedMethods.size / 5.0).toInt()
            message.edit { it.setEmbed { it.buildMessage(sortedMethods, mappingsContainer, page, user, maxPage, isYarn) } }.subscribe { msg ->
                if (channel.type.name.startsWith("GUILD_"))
                    msg.removeAllReactions().block()
                msg.subscribeReactions("⬅", "❌", "➡")
                api.eventDispatcher.on(ReactionAddEvent::class.java).filter { e -> e.messageId == msg.id }.take(Duration.ofMinutes(15)).subscribe {
                    when (it.userId) {
                        api.selfId.get() -> {
                        }
                        user.id -> {
                            if (!it.emoji.asUnicodeEmoji().isPresent) {
                                msg.removeReaction(it.emoji, it.userId).subscribe()
                            } else {
                                val unicode = it.emoji.asUnicodeEmoji().get()
                                if (unicode.raw == "❌") {
                                    msg.delete().subscribe()
                                } else if (unicode.raw == "⬅") {
                                    msg.removeReaction(it.emoji, it.userId).subscribe()
                                    if (page > 0) {
                                        page--
                                        msg.edit { it.setEmbed { it.buildMessage(sortedMethods, mappingsContainer, page, user, maxPage, isYarn) } }.subscribe()
                                    }
                                } else if (unicode.raw == "➡") {
                                    msg.removeReaction(it.emoji, it.userId).subscribe()
                                    if (page < maxPage - 1) {
                                        page++
                                        msg.edit { it.setEmbed { it.buildMessage(sortedMethods, mappingsContainer, page, user, maxPage, isYarn) } }.subscribe()
                                    }
                                } else {
                                    msg.removeReaction(it.emoji, it.userId).subscribe()
                                }
                            }
                        }
                        else -> msg.removeReaction(it.emoji, it.userId).subscribe()
                    }
                }
            }
        } catch (t: Throwable) {
            try {
                message.edit { it.setEmbed { it.generateThrowable(t, user) } }.subscribe()
            } catch (throwable2: Throwable) {
                throwable2.addSuppressed(t)
                throw throwable2
            }
        }
    }
}

private enum class FindMethodMethod {
    INTERMEDIARY,
    MAPPED,
    OBF_CLIENT,
    OBF_SERVER,
    OBF_MERGED,
    WILDCARD
}

private data class MethodWrapper(val method: Method, val parent: Class, val cm: FindMethodMethod)

private fun EmbedCreateSpec.buildMessage(sortedMethods: List<MethodWrapper>, mappingsContainer: MappingsContainer, page: Int, author: User, maxPage: Int, isYarn: Boolean) {
    setFooter("Requested by " + author.discriminatedName, author.avatarUrl)
    setTimestampToNow()
    if (maxPage > 1) setTitle("List of ${mappingsContainer.name} Mappings (Page ${page + 1}/$maxPage)")
    var desc = ""
    sortedMethods.dropAndTake(5 * page, 5).forEach {
        if (!desc.isEmpty())
            desc += "\n\n"
        val obfMap = LinkedHashMap<String, String>()
        if (!it.method.obfName.isMerged()) {
            if (it.method.obfName.client != null) obfMap["client"] = it.method.obfName.client!!
            if (it.method.obfName.server != null) obfMap["server"] = it.method.obfName.server!!
        }
        desc += "**MC ${mappingsContainer.version}: ${it.parent.mappedName
                ?: it.parent.intermediaryName}.${it.method.mappedName ?: it.method.intermediaryName}**\n" +
                "__Name__: " + (if (it.method.obfName.isEmpty()) "" else if (it.method.obfName.isMerged()) "${it.method.obfName.merged} => " else "${obfMap.entries.joinToString { "${it.key}=**${it.value}**" }} => ") +
                "`${it.method.intermediaryName}`" + (if (it.method.mappedName == null || it.method.mappedName == it.method.intermediaryName) "" else " => `${it.method.mappedName}`")
//        desc += "\n__Descriptor__: `${if (it.method.mappedName == null) it.method.intermediaryName else it.method.mappedName}${it.method.mappedDesc
//                ?: it.method.intermediaryDesc.mapIntermediaryDescToNamed(mappingsContainer)}`"
        if (isYarn) {
            desc += "\n__Mixin Target__: `L${it.parent.mappedName
                    ?: it.parent.intermediaryName};${if (it.method.mappedName == null) it.method.intermediaryName else it.method.mappedName}${it.method.mappedDesc
                    ?: it.method.intermediaryDesc.mapIntermediaryDescToNamed(mappingsContainer)}`"
        } else {
            desc += "\n__AT__: `public ${(it.parent.intermediaryName).replace('/', '.')}" +
                    " ${it.method.intermediaryName}${it.method.obfDesc.merged!!.mapObfDescToNamed(mappingsContainer)}" +
                    " # ${if (it.method.mappedName == null) it.method.intermediaryName else it.method.mappedName}`"
        }
    }
    setDescription(desc.substring(0, min(desc.length, 2000)))
}

private fun String.mapObfDescToNamed(container: MappingsContainer): String {
    if (startsWith('(') && contains(')')) {
        val split = split(')')
        val parametersOG = split[0].substring(1).toCharArray()
        val returnsOG = split[1].toCharArray()
        val parametersUnmapped = mutableListOf<String>()
        val returnsUnmapped = mutableListOf<String>()

        var lastT: String? = null
        for (char in parametersOG) {
            when {
                lastT != null && char == ';' -> {
                    parametersUnmapped.add(lastT)
                    lastT = null
                }
                lastT != null -> {
                    lastT += char
                }
                char == 'L' -> {
                    lastT = ""
                }
                else -> parametersUnmapped.add(char.toString())
            }
        }
        for (char in returnsOG) {
            when {
                lastT != null && char == ';' -> {
                    returnsUnmapped.add(lastT)
                    lastT = null
                }
                lastT != null -> {
                    lastT += char
                }
                char == 'L' -> {
                    lastT = ""
                }
                else -> returnsUnmapped.add(char.toString())
            }
        }
        return "(" + parametersUnmapped.joinToString("") {
            if (it.length != 1) {
                "L${container.getClassByObfName(it)?.intermediaryName ?: it};"
            } else
                it
        } + ")" + returnsUnmapped.joinToString("") {
            if (it.length != 1) {
                "L${container.getClassByObfName(it)?.intermediaryName ?: it};"
            } else
                it
        }
    }
    return this
}
